In [1]:
# Supress Warnings

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

## Task 1: Reading and Inspection

-  ### Subtask 1.1: Import and read

Import and read the movie database. Store it in a variable called `movies`.

In [3]:
movies = pd.read_csv("Movie+Assignment+Data.csv")# Write your code for importing the csv file here.

-  ### Subtask 1.2: Inspect the dataframe

Inspect the dataframe's columns, shapes, variable types etc.

In [4]:
# Using shape attribute,save number of rows & columns in variables 
# 'orig_num_rows','orig_num_cols' which may be required for future analysis.
orig_num_rows , orig_num_cols = movies.shape 

# Print the number of Rows and Columns
print("The 'movies' dataframe has {} Rows and {} Columns.".format(orig_num_rows,orig_num_cols))

The 'movies' dataframe has 5043 Rows and 28 Columns.


In [5]:
# Write your code for inspection here
movies.dtypes # To understand the various attributes and their type in the movies dataframe.

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

High number of null values in **'gross'** and **'budget'** columns.

In [6]:
movies.head() # Use head attribute to display the first 5 rows of the dataframe to get a feel of the dataframe layout.

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                 40000.0  309404152.0         Action|Adventure|Fantasy   
2                 11000.0  200074175.0        Action|Adventure|Thriller   
3                 27000.0  448130642.0                  Action|Thriller   
4                   131.0          NaN                      Documentary   

          ...          num_user_for_reviews language  country  content_rating  \
0         ...                        3054.0  English      USA           PG-13   
1         ...                        1238.0  English      USA           PG-13   
2         ...                         994.0  English       UK           PG-13   
3         ...                        2701.0  English      USA           PG-13   
4         ...                           NaN      NaN      NaN             NaN   

        budget  title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0  237000000.0      2009.0                  936.0        7.9          1.78   
1  300000000.0      2007.0                 5000.0        7.1          2.35   
2  245000000.0      2015.0                  393.0        6.8          2.35   
3  250000000.0      2012.0                23000.0        8.5          2.35   
4          NaN         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [7]:
# Column-wise null percentages to get a feel of missing data
print("Column-wise null percentages to get a feel of missing data : \n{}" \
      .format(round(100*(movies.isnull().sum()/len(movies.index)), 2)))

Column-wise null percentages to get a feel of missing data : 
color                         0.38
director_name                 2.06
num_critic_for_reviews        0.99
duration                      0.30
director_facebook_likes       2.06
actor_3_facebook_likes        0.46
actor_2_name                  0.26
actor_1_facebook_likes        0.14
gross                        17.53
genres                        0.00
actor_1_name                  0.14
movie_title                   0.00
num_voted_users               0.00
cast_total_facebook_likes     0.00
actor_3_name                  0.46
facenumber_in_poster          0.26
plot_keywords                 3.03
movie_imdb_link               0.00
num_user_for_reviews          0.42
language                      0.24
country                       0.10
content_rating                6.01
budget                        9.76
title_year                    2.14
actor_2_facebook_likes        0.26
imdb_score                    0.00
aspect_ratio                

High percentage of null values in **'gross'** and **'budget'** columns.

## Task 2: Cleaning the Data

-  ### Subtask 2.1: Inspect Null values

Find out the number of Null values in all the columns and rows. Also, find the percentage of Null values in each column. Round off the percentages upto two decimal places.

In [8]:
# Write your code for column-wise null count here.
movies.isnull().sum() # isnull() with sum() is used to get a column-wise null count.

color                         19
director_name                104
num_critic_for_reviews        50
duration                      15
director_facebook_likes      104
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross                        884
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                153
movie_imdb_link                0
num_user_for_reviews          21
language                      12
country                        5
content_rating               303
budget                       492
title_year                   108
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 329
movie_facebook_likes           0
dtype: int64

#### Top 2 columns by number of missing values

**884** rows have missing values in **'gross'** column and **492** rows have missing values in **'budget'** column.

In [9]:
# Write your code for row-wise null count here
movies.isnull().sum(axis=1)  # axis is set to 1 to get row-wise null count.

0        0
1        0
2        0
3        0
4       14
5        0
6        0
7        0
8        0
9        0
10       0
11       0
12       0
13       0
14       0
15       0
16       0
17       0
18       0
19       0
20       0
21       0
22       0
23       0
24       0
25       0
26       0
27       0
28       0
29       0
        ..
5013     2
5014     2
5015     0
5016     5
5017     2
5018     2
5019     2
5020     5
5021     1
5022     4
5023     2
5024     1
5025     0
5026     0
5027     0
5028     3
5029     1
5030     5
5031     3
5032     4
5033     0
5034     1
5035     0
5036     3
5037     1
5038     4
5039     5
5040     4
5041     2
5042     0
Length: 5043, dtype: int64

In [10]:
# Write your code for column-wise null percentages here
# (1). movies.isnull().sum() : gives column-wise null count.
# (2). len(movies.index) : gives number of rows.
# Dividing (1) by (2) and multiplying by 100 to get percentage.Rounding off to 2-decimal places using round(..,2).
print(round(100*(movies.isnull().sum()/len(movies.index)), 2))

color                         0.38
director_name                 2.06
num_critic_for_reviews        0.99
duration                      0.30
director_facebook_likes       2.06
actor_3_facebook_likes        0.46
actor_2_name                  0.26
actor_1_facebook_likes        0.14
gross                        17.53
genres                        0.00
actor_1_name                  0.14
movie_title                   0.00
num_voted_users               0.00
cast_total_facebook_likes     0.00
actor_3_name                  0.46
facenumber_in_poster          0.26
plot_keywords                 3.03
movie_imdb_link               0.00
num_user_for_reviews          0.42
language                      0.24
country                       0.10
content_rating                6.01
budget                        9.76
title_year                    2.14
actor_2_facebook_likes        0.26
imdb_score                    0.00
aspect_ratio                  6.52
movie_facebook_likes          0.00
dtype: float64


#### Top 2 columns by percentage of missing values
The columns **'gross'** has **17.53%** missing values and column **budget** has **9.76%** missing values.

-  ### Subtask 2.2: Drop unecessary columns

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [11]:
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations 

# Print the original list of columns in the movies dataframe
print("\nOriginal List of Columns :")
print(list(movies.columns))

# Print the original number of columns in the movies dataframe using the 'orig_num_cols' variable.
print("\nOriginal Number of columns : {}.\n".format(orig_num_cols))

# Column-wise null percentages before dropping unwanted columns
print("Column-wise null percentages before dropping unwanted columns : \n{}" \
      .format(round(100*(movies.isnull().sum()/len(movies.index)), 2)))

# Number of rows before dropping unwanted columns
print("\nNumber of rows before dropping unwanted columns: {}".format(len(movies.index)))


Original List of Columns :
['color', 'director_name', 'num_critic_for_reviews', 'duration', 'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name', 'movie_title', 'num_voted_users', 'cast_total_facebook_likes', 'actor_3_name', 'facenumber_in_poster', 'plot_keywords', 'movie_imdb_link', 'num_user_for_reviews', 'language', 'country', 'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes', 'imdb_score', 'aspect_ratio', 'movie_facebook_likes']

Original Number of columns : 28.

Column-wise null percentages before dropping unwanted columns : 
color                         0.38
director_name                 2.06
num_critic_for_reviews        0.99
duration                      0.30
director_facebook_likes       2.06
actor_3_facebook_likes        0.46
actor_2_name                  0.26
actor_1_facebook_likes        0.14
gross                        17.53
genres                        0.00
actor_1_name       

In [12]:
# Based on inspection, following is the list of columns to be dropped from the 'movies' dataframe
cols_to_drop = ['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes', 
                 'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes', 'actor_3_name', 
                 'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link',
                 'aspect_ratio', 'plot_keywords']

print("\n Dropping columns starts... \n")

# Use Dataframe.drop() function to drop unwanted columns present in the lis 'cols_to_drop'
# columns=cols_to_drop : Passes the list of columns to drop
# axis=1 : To indicate that we intend to drop the entire columns
# inplace=True : To modify the movies dataframe itself i.e., drop columns from 'movies' dataframe itself.
movies.drop(columns=cols_to_drop, axis=1, inplace=True)

print("\n Dropping columns ends... \n")


 Dropping columns starts... 


 Dropping columns ends... 



In [13]:
# Print the list of columns left in the 'movies' dataframe after dropping unwanted columns
print("List of Columns left after dropping unwanted columns :")
print(list(movies.columns))

# Print the number of columns left in the 'movies' dataframe after dropping unwanted columns
print("\nNumber of columns left after dropping unwanted columns : {}.\n".format(movies.columns.size))

# Column-wise null percentages after dropping unwanted columns
print("Column-wise null percentages after dropping unwanted columns : \n{}" \
      .format(round(100*(movies.isnull().sum()/len(movies.index)), 2)))

# Number of rows after dropping unwanted columns
print("\nNumber of rows after dropping unwanted columns: {}".format(len(movies.index)))

List of Columns left after dropping unwanted columns :
['director_name', 'num_critic_for_reviews', 'gross', 'genres', 'actor_1_name', 'movie_title', 'num_voted_users', 'num_user_for_reviews', 'language', 'budget', 'title_year', 'imdb_score', 'movie_facebook_likes']

Number of columns left after dropping unwanted columns : 13.

Column-wise null percentages after dropping unwanted columns : 
director_name              2.06
num_critic_for_reviews     0.99
gross                     17.53
genres                     0.00
actor_1_name               0.14
movie_title                0.00
num_voted_users            0.00
num_user_for_reviews       0.42
language                   0.24
budget                     9.76
title_year                 2.14
imdb_score                 0.00
movie_facebook_likes       0.00
dtype: float64

Number of rows after dropping unwanted columns: 5043


-  ### Subtask 2.3: Drop unecessary rows using columns with high Null percentages

Now, on inspection you might notice that some columns have large percentage (greater than 5%) of Null values. Drop all the rows which have Null values for such columns.

In [14]:
# Check Column-wise null percentages
print("Column-wise null percentages :")
print(round(100*(movies.isnull().sum()/len(movies.index)), 2))

Column-wise null percentages :
director_name              2.06
num_critic_for_reviews     0.99
gross                     17.53
genres                     0.00
actor_1_name               0.14
movie_title                0.00
num_voted_users            0.00
num_user_for_reviews       0.42
language                   0.24
budget                     9.76
title_year                 2.14
imdb_score                 0.00
movie_facebook_likes       0.00
dtype: float64


The following columns have greater than 5% of Null Values :

**gross :                     17.53%**

**budget :                     9.76%**

Drop rows which have null values for columns **'gross'** and **'budget'**.

In [15]:
# Write your code for dropping the rows here

# Remove rows which has Null Value for column 'gross'
movies = movies[~np.isnan(movies['gross'])]

# Remove rows which has Null Value for column 'budget'
movies = movies[~np.isnan(movies['budget'])]

# Check Column-wise null percentages after dropping unwanted rows having null-value in 'gross' and budget' columns
print("\nColumn-wise null percentages after dropping rows having null-value in 'gross' and 'budget' columns: \n ")
print(round(100*(movies.isnull().sum()/len(movies.index)), 2))


Column-wise null percentages after dropping rows having null-value in 'gross' and 'budget' columns: 
 
director_name             0.00
num_critic_for_reviews    0.03
gross                     0.00
genres                    0.00
actor_1_name              0.08
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
language                  0.08
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64


In [16]:
#Original Number of rows
print("\nOriginal number of rows : {}".format(orig_num_rows))

# Number of rows retained at end of Subtask 2.4
print("\nNumber of rows retained at end of Subtask 2.3: {}".format(len(movies.index)))

# Percentage of rows retained at end of Subtask 2.3
print("\nPercentage of rows retained at end of Subtask 2.3: {}%".format(round(100*(len(movies.index)/orig_num_rows), 2)))


Original number of rows : 5043

Number of rows retained at end of Subtask 2.3: 3891

Percentage of rows retained at end of Subtask 2.3: 77.16%


-  ### Subtask 2.4: Drop unecessary rows

Some of the rows might have greater than five NaN values. Such rows aren't of much use for the analysis and hence, should be removed.

In [17]:
# Checking rows which have more than 5 NaN values
print("Number of rows with more than 5 NaN values: {}.".format(sum(movies.isnull().sum(axis=1) > 5)))

# Checking rows which have atleast 1 NaN value
print("Number of rows with atleast 1 NaN value: {}.".format(sum(movies.isnull().sum(axis=1) > 0)))

Number of rows with more than 5 NaN values: 0.
Number of rows with atleast 1 NaN value: 7.


In [18]:
# Although there are no rows which have more than 5 NaN Values, 
# still going ahead with the task to remove rows with more than 5 NaN values
# for illustrative purpose.

# Write your code for dropping the rows here
movies = movies[movies.isnull().sum(axis=1) <= 5]

In [19]:
#Original Number of rows
print("\nOriginal number of rows : {}".format(orig_num_rows))

# Number of rows retained at end of Subtask 2.4
print("\nNumber of rows retained at end of Subtask 2.4: {}".format(len(movies.index)))

# Percentage of rows retained at end of Subtask 2.4
print("\nPercentage of rows retained at end of Subtask 2.4: {}%".format(round(100*(len(movies.index)/orig_num_rows), 2)))


Original number of rows : 5043

Number of rows retained at end of Subtask 2.4: 3891

Percentage of rows retained at end of Subtask 2.4: 77.16%


-  ### Subtask 2.5: Fill NaN values

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [20]:
# Using value_counts() and head(10) functions to get the Top 10 languages in 'language' column
print("\nTop 10 Languages in 'language' column:\n{}".format(movies.language.value_counts().head(10)) )

# Print total number of rows in the dataframe
print("\nNumber of rows in 'movies' dataframe:",len(movies.index))


Top 10 Languages in 'language' column:
English       3707
French          37
Spanish         26
Mandarin        15
German          13
Japanese        12
Hindi           10
Cantonese        8
Italian          7
Portuguese       5
Name: language, dtype: int64

Number of rows in 'movies' dataframe: 3891


Of the 3891 movies left in the 'movies' dataframe, 3707 movies are in **'English'** language.

**'English'** emerges as the top language in the movies dataframe.

In [21]:
# Write your code for filling the NaN values in the 'language' column here
# Using fillna() function, 'English' is imputed in place of NaN values
# inplace=True : Modifies the movies database itself (inplace operation)
movies.language.fillna('English' , inplace=True)

# The above line of code can also be written as 
# movies.loc[pd.isnull(movies['language']), ['language']] = 'English' or
#movies.loc[pd.isnull(movies['language']), ['language']] = movies.language.describe().top 

In [22]:
# Using value_counts() and head(10) functions to get the Top 10 languages in 'language' column
print("\nTop 10 Languages in 'language' column after imputing:\n{}".format(movies.language.value_counts().head(10)) )


Top 10 Languages in 'language' column after imputing:
English       3710
French          37
Spanish         26
Mandarin        15
German          13
Japanese        12
Hindi           10
Cantonese        8
Italian          7
Portuguese       5
Name: language, dtype: int64


The number of records with 'English' language has increased from 3707 to 3710 , meaning 3 rows with null values in 'language' column have been filled with 'English'.

-  ### Subtask 2.6: Check the number of retained rows

You might notice that two of the columns viz. `num_critic_for_reviews` and `actor_1_name` have small percentages of NaN values left. You can let these columns as it is for now. Check the number and percentage of the rows retained after completing all the tasks above.

In [23]:
# Check Column-wise null percentages to confirm that there are still 
# some null values present in num_critic_for_reviews and actor_1_name columns      
print("\nColumn-wise null percentages Subtask 2.6 :")
print(round(100*(movies.isnull().sum()/len(movies.index)), 2))


Column-wise null percentages Subtask 2.6 :
director_name             0.00
num_critic_for_reviews    0.03
gross                     0.00
genres                    0.00
actor_1_name              0.08
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
language                  0.00
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64


Columns 'num_critic_for_reviews' and 'actor_1_name' indeed have a small percentage of NaN Values.

In [24]:
# Write your code for checking number of retained rows here

#Original Number of rows
print("\nOriginal number of rows : {}".format(orig_num_rows))

num_rows_retained = len(movies.index) #Number of rows retained stored in 'num_rows_retained' for later use
print("\nNumber of rows retained : {}".format(num_rows_retained))

# Percentage of rows retained
print("\nPercentage of rows retained : {}%".format(round(100*(num_rows_retained/orig_num_rows), 2)))


Original number of rows : 5043

Number of rows retained : 3891

Percentage of rows retained : 77.16%


**Checkpoint 1:** You might have noticed that we still have around `77%` of the rows! 

## Task 3: Data Analysis

-  ### Subtask 3.1: Change the unit of columns

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [25]:
# Write your code for unit conversion here
# 1 million = 1000000. 
# Divide by 1000000 to convert unit of 'budget' and 'gross' columns to million $ and assign back to same columns.
movies['budget'] = movies['budget']/1000000
movies['gross'] = movies['gross']/1000000

# Note: Since no instruction has been given to round off the values in 'budget' and 'gross' columns, the values are left as is.
# Below code for converting $ to million $ and rounding off to 2 decimal places
#movies['budget'] = round((movies['budget']/1000000),2)
#movies['gross'] = round((movies['gross']/1000000),2)

In [26]:
# Display 'budget' and 'gross' columns for first 5 rows to check the conversion from $ to million $
movies[['budget','gross']].head()

budget       gross
0   237.0  760.505847
1   300.0  309.404152
2   245.0  200.074175
3   250.0  448.130642
5   263.7   73.058679

-  ### Subtask 3.2: Find the movies with highest profit

    1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
    2. Sort the dataframe using the `profit` column as reference.
    3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [27]:
# Write your code for creating the profit column here
# Calculation: movie profit = gross collections - movie budget
# Subtract column 'budget' from column 'gross' and assign it to a new column 'profit'
movies['profit'] = movies['gross'] - movies['budget'] 

# The same thing as above can be done using
# movies['profit'] = movies.gross - movies.budget

In [28]:
# Check first 5 rows of the dataframe to confirm the creation of the new column 'profit'
movies.head()

director_name  num_critic_for_reviews       gross  \
0      James Cameron                   723.0  760.505847   
1     Gore Verbinski                   302.0  309.404152   
2         Sam Mendes                   602.0  200.074175   
3  Christopher Nolan                   813.0  448.130642   
5     Andrew Stanton                   462.0   73.058679   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
5          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
5                               John Carter            212204   

   num_user_for_reviews language  budget  title_year  imdb_score  \
0                3054.0  English   237.0      2009.0         7.9   
1                1238.0  English   300.0      2007.0         7.1   
2                 994.0  English   245.0      2015.0         6.8   
3                2701.0  English   250.0      2012.0         8.5   
5                 738.0  English   263.7      2012.0         6.6   

   movie_facebook_likes      profit  
0                 33000  523.505847  
1                     0    9.404152  
2                 85000  -44.925825  
3                164000  198.130642  
5                 24000 -190.641321

In [29]:
# Write your code for sorting the dataframe here
# sort_values() function is used to sort the dataframe 
# by='profit' : indicates that the sorting needs to be done based on 'profit' column
# ascending=False : indicates that the sorting needs to be done in descending order
# Using the argument inplace=True to sort on 'profit' column and modify the movies dataframes itself (inplace operation).
movies.sort_values(by='profit' , ascending=False , inplace=True)

Please note that an inplace sorting was done on the movies database as the next task (3.2.3) states :

**"3. Extract the top ten profiting movies in descending order and store them in a new dataframe - top10"**

which implies that the 'movies' database is already sorted on the 'profit' column and we just extract the top 10 movies from it.

In [30]:
# Write your code to get the top 10 profiting movies here

# Get Top ten rows using head(10)
top10 = movies.head(10) # Alternate way of achieving the same : top10 = movies[:10]

-  ### Subtask 3.3: Drop duplicate values

After you found out the top 10 profiting movies, you might have notice a duplicate value. So, it seems like the dataframe has duplicate values as well. Drop the duplicate values from the dataframe and repeat `Subtask 3.2`.

In [31]:
# Display top10 dataframe to check for duplicate rows
top10

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
29      Colin Trevorrow                   644.0  652.177271   
26        James Cameron                   315.0  658.672302   
3024       George Lucas                   282.0  460.935665   
3080   Steven Spielberg                   215.0  434.949459   
794         Joss Whedon                   703.0  623.279547   
17          Joss Whedon                   703.0  623.279547   
509        Roger Allers                   186.0  422.783777   
240        George Lucas                   320.0  474.544677   
66    Christopher Nolan                   645.0  533.316061   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                               Drama|Romance    Leonardo DiCaprio   
3024           Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                             Family|Sci-Fi         Henry Thomas   
794                    Action|Adventure|Sci-Fi      Chris Hemsworth   
17                     Action|Adventure|Sci-Fi      Chris Hemsworth   
509   Adventure|Animation|Drama|Family|Musical    Matthew Broderick   
240            Action|Adventure|Fantasy|Sci-Fi      Natalie Portman   
66                 Action|Crime|Drama|Thriller       Christian Bale   

                                     movie_title  num_voted_users  \
0                                        Avatar            886204   
29                               Jurassic World            418214   
26                                      Titanic            793059   
3024         Star Wars: Episode IV - A New Hope            911097   
3080                 E.T. the Extra-Terrestrial            281842   
794                                The Avengers            995415   
17                                 The Avengers            995415   
509                               The Lion King            644348   
240   Star Wars: Episode I - The Phantom Menace            534658   
66                              The Dark Knight           1676169   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
29                  1290.0  English   150.0      2015.0         7.0   
26                  2528.0  English   200.0      1997.0         7.7   
3024                1470.0  English    11.0      1977.0         8.7   
3080                 515.0  English    10.5      1982.0         7.9   
794                 1722.0  English   220.0      2012.0         8.1   
17                  1722.0  English   220.0      2012.0         8.1   
509                  656.0  English    45.0      1994.0         8.5   
240                 3597.0  English   115.0      1999.0         6.5   
66                  4667.0  English   185.0      2008.0         9.0   

      movie_facebook_likes      profit  
0                    33000  523.505847  
29                  150000  502.177271  
26                   26000  458.672302  
3024                 33000  449.935665  
3080                 34000  424.449459  
794                 123000  403.279547  
17                  123000  403.279547  
509                  17000  377.783777  
240                  13000  359.544677  
66                   37000  348.316061

The movie **'The Avengers'** appears 2 times(on rows 6 and 7) in the top10 dataframe, indicating that there are duplicate rows present in the movies dataframe 

In [32]:
# Using the duplicated() function, find the number of duplicate rows

print("Number of duplicate rows: {}.\n".format(movies.duplicated().sum()))

# Display the duplicate rows, which can be later used for validation if necessary.

movies[movies.duplicated()]

Number of duplicate rows: 35.



director_name  num_critic_for_reviews       gross  \
17             Joss Whedon                   703.0  623.279547   
174            Bill Condon                   322.0  292.298923   
395              Rob Cohen                   187.0  144.512310   
2025          Herbert Ross                    60.0   80.000000   
656         Paul Verhoeven                   196.0  119.412921   
663    Angelina Jolie Pitt                   322.0  115.603980   
3193           Paul Haggis                   287.0   54.557348   
2619        John Carpenter                   318.0   47.000000   
3984        Peter Cattaneo                   122.0   45.857453   
2181             Jon Lucas                    81.0   55.461307   
1984          Ole Bornedal                   264.0   49.122319   
2628            Shawn Levy                    69.0   47.811275   
2935           Tamra Davis                   111.0   37.188667   
3017           Guy Ritchie                   151.0   30.093107   
2533           Neil Burger                   236.0   39.825798   
2777        Stephen Frears                    51.0   34.700000   
4182            Rob Zombie                   220.0   33.386128   
3064      Tim Blake Nelson                    92.0   16.017403   
2134              Frank Oz                   168.0    8.579684   
2568         Vic Armstrong                   169.0   13.998282   
1096         Albert Hughes                   208.0   31.598308   
1962            Wes Craven                   160.0   14.637490   
2063           Andy Cadiff                    65.0   12.189514   
4313       Bruce McCulloch                    52.0   13.973532   
2454       Kenneth Branagh                    85.0    4.414535   
1697         Paul McGuigan                    98.0   12.831121   
2584           Danny Boyle                   393.0    2.319187   
3480  Michael Winterbottom                    71.0    0.403932   
314           Brett Ratner                   245.0   72.660029   
1171           Yimou Zhang                   283.0    0.084961   
1146         Paul McGuigan                   159.0    5.773519   
63             David Yates                   248.0  124.051759   
389             Josh Trank                   369.0   56.114221   
2971      John Lee Hancock                   106.0   22.406362   
303             Joe Wright                   256.0   34.964818   

                                     genres          actor_1_name  \
17                  Action|Adventure|Sci-Fi       Chris Hemsworth   
174         Adventure|Drama|Fantasy|Romance      Robert Pattinson   
395                   Action|Crime|Thriller           Paul Walker   
2025                    Drama|Music|Romance          Dianne Wiest   
656                           Action|Sci-Fi             Ronny Cox   
663               Biography|Drama|Sport|War         Finn Wittrock   
3193                   Crime|Drama|Thriller           Don Cheadle   
2619                        Horror|Thriller      Jamie Lee Curtis   
3984                     Comedy|Drama|Music         Tom Wilkinson   
2181                                 Comedy            Mila Kunis   
1984                        Horror|Thriller         Kyra Sedgwick   
2628                Adventure|Comedy|Family         Frankie Muniz   
2935                           Comedy|Drama        Britney Spears   
3017                           Comedy|Crime         Jason Statham   
2533         Drama|Mystery|Romance|Thriller          Rufus Sewell   
2777                          Drama|Romance          Keanu Reeves   
4182                                 Horror  Scout Taylor-Compton   
3064                 Drama|Romance|Thriller          Mekhi Phifer   
2134                                 Comedy        Peter Dinklage   
2568  Action|Drama|Fantasy|Mystery|Thriller          Nicolas Cage   
1096                Horror|Mystery|Thriller           Johnny Depp   
1962                Horror|Mystery|Thriller          Frank Grillo   
2063                         Comedy|Romance     A

The above analysis shows that there are in all **35 duplicate rows** that need to be deleted from the movies dataframe.

In [33]:
# Write your code for dropping duplicate values here
# Remove duplicate using drop_duplicates() function
# inplace=True : to remove duplicates from movies dataframe itself(inplace operation)

movies.drop_duplicates(inplace=True) #Check any other way to remove duplicates

In [34]:
# Number of rows before removing duplicate rows
# Got from variable 'num_rows_retained' saved under SubTask 2.6
print("\nNumber of rows before removing duplicate rows     : {}".format(num_rows_retained))

# Number of rows left after removing duplicate rows
num_rows_unique = len(movies.index) #variable 'num_rows_unique' to save the number of rows left after removing duplicates
print("\nNumber of rows left after removing duplicate rows : {}".format(len(movies.index)))

#Number of duplicate rows removed
print("\nNumber of duplicate rows removed : {}".format(num_rows_retained - len(movies.index)))

# Percentage of rows retained
print("\nPercentage of rows retained : {}%".format(round(100*(len(movies.index)/orig_num_rows), 2)))


Number of rows before removing duplicate rows     : 3891

Number of rows left after removing duplicate rows : 3856

Number of duplicate rows removed : 35

Percentage of rows retained : 76.46%


In [35]:
# Write code for repeating subtask 2 here
# Creating the profit column here - 3.2.1 -- Skipping this step as this is redundant. 
# Column 'profit' has already been added to the 'movies' dataframe.
#movies['profit'] = movies['gross'] - movies['budget'] 

# Sorting the dataframe here on 'profit' column - 3.2.2 -- Skipping this step as this is redundant
# 'movies' dataframe already sorted by 'profit' column in descending order.
#movies.sort_values(by='profit' , ascending=False , inplace=True) 

# Get the top 10 profiting movies - 3.2.3
top10 = movies.head(10) 

# Display top10 dataframe 
top10

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
29      Colin Trevorrow                   644.0  652.177271   
26        James Cameron                   315.0  658.672302   
3024       George Lucas                   282.0  460.935665   
3080   Steven Spielberg                   215.0  434.949459   
794         Joss Whedon                   703.0  623.279547   
509        Roger Allers                   186.0  422.783777   
240        George Lucas                   320.0  474.544677   
66    Christopher Nolan                   645.0  533.316061   
439           Gary Ross                   673.0  407.999255   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                               Drama|Romance    Leonardo DiCaprio   
3024           Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                             Family|Sci-Fi         Henry Thomas   
794                    Action|Adventure|Sci-Fi      Chris Hemsworth   
509   Adventure|Animation|Drama|Family|Musical    Matthew Broderick   
240            Action|Adventure|Fantasy|Sci-Fi      Natalie Portman   
66                 Action|Crime|Drama|Thriller       Christian Bale   
439            Adventure|Drama|Sci-Fi|Thriller    Jennifer Lawrence   

                                     movie_title  num_voted_users  \
0                                        Avatar            886204   
29                               Jurassic World            418214   
26                                      Titanic            793059   
3024         Star Wars: Episode IV - A New Hope            911097   
3080                 E.T. the Extra-Terrestrial            281842   
794                                The Avengers            995415   
509                               The Lion King            644348   
240   Star Wars: Episode I - The Phantom Menace            534658   
66                              The Dark Knight           1676169   
439                            The Hunger Games            701607   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
29                  1290.0  English   150.0      2015.0         7.0   
26                  2528.0  English   200.0      1997.0         7.7   
3024                1470.0  English    11.0      1977.0         8.7   
3080                 515.0  English    10.5      1982.0         7.9   
794                 1722.0  English   220.0      2012.0         8.1   
509                  656.0  English    45.0      1994.0         8.5   
240                 3597.0  English   115.0      1999.0         6.5   
66                  4667.0  English   185.0      2008.0         9.0   
439                 1959.0  English    78.0      2012.0         7.3   

      movie_facebook_likes      profit  
0                    33000  523.505847  
29                  150000  502.177271  
26                   26000  458.672302  
3024                 33000  449.935665  
3080                 34000  424.449459  
794                 123000  403.279547  
509                  17000  377.783777  
240                  13000  359.544677  
66                   37000  348.316061  
439                 140000  329.999255

**James Cameron**'s movies **'Avatar'** and **'Titanic'** at rows 1 and 3.

The movie **'The Avengers'** which had a duplicate row now appears **only once** on rows 6.

**Checkpoint 2:** You might spot two movies directed by `James Cameron` in the list.

---

#### Additional Analysis
On closer inspection of the dataset, noticed that there are a few more rows which have the same values in all the columns except for the 'num_voted_users' column. For example, consider 'movie_title' = 'Pan' which has two entries left after removal of duplicates. Note that the difference betweeen the two rows is in the coulmn 'num_voted_users', which I think is due to some error and one of the rows has to be removed for correct analysis. However, as per a reply by TA on the discussion forum, we have been advised to not drop these rows. Also, these rows affect the outcome of analysis w.r.t IMBb_Top_250 later in the assignment.

In case, we need to handle such situations, we can use the subset argument and pass 'movie_title' as the column based on which the duplicates can be removed. Please find below the commented code for such an analysis where duplicates are removed based on 'movie_title' field by keeping only the first occurence and deleting all others.

In [36]:
# Display Movies with movie_title 'Pan' starring 'Hugh Jackman'
movies[movies['movie_title'].str.contains('Pan') & movies['actor_1_name'].str.contains('Hugh Jackman')]

director_name  num_critic_for_reviews      gross  \
3879    Joe Wright                   256.0  34.964818   
145     Joe Wright                   256.0  34.964818   

                        genres  actor_1_name movie_title  num_voted_users  \
3879  Adventure|Family|Fantasy  Hugh Jackman        Pan             39975   
145   Adventure|Family|Fantasy  Hugh Jackman        Pan             39956   

      num_user_for_reviews language  budget  title_year  imdb_score  \
3879                 186.0  English   150.0      2015.0         5.8   
145                  186.0  English   150.0      2015.0         5.8   

      movie_facebook_likes      profit  
3879                 24000 -115.035182  
145                  24000 -115.035182

In [37]:
# Using the duplicated() function, find the numbe of duplicate rows
# subset='movie_title' : find duplicate rows based on column 'movie_title'
# keep='fisrt' : keeps only the first row and marks other rows as duplicate rows 

#print("Number of duplicate rows: {}.\n".format(movies.duplicated(subset='movie_title' , keep='first').sum()))

# Print the duplicate rows, which can be later used for validation if necessary.

#movies[movies.duplicated(subset='movie_title' , keep='first')]

#The above command will remove 102 duplicate rows from the movies dataframe.

# Write your code for dropping duplicate values here
# Remove duplicate using drop_duplicates() function
# subset='movie_title' : to check for duplicates based on column 'movie_title' only
# keep='first' : to keep only the first occurence and remove all other duplicate rows from the movies dataframe
# inplace=True : to remove duplicates from movies dataframe itself(inplace operation)

#movies.drop_duplicates(subset='movie_title',keep='first',inplace=True) #Check any other way to remove duplicates

---

-  ### Subtask 3.4: Find IMDb Top 250

    1. Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000.
Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.
    2. Extract all the movies in the `IMDb_Top_250` dataframe which are not in the English language and store them in a new dataframe named `Top_Foreign_Lang_Film`.

In [38]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'

# movies.loc[movies.num_voted_users > 25000 ] : Chooses only those rows whose 'num_voted_users' value is greater than 25000
# sort_values(by='imdb_score' , ascending=False) : sorts the dataframe by column 'imdb_score' in descending order
# head(250) : Chooses the first 250 rows from the dataframe.

IMDb_Top_250 = (movies.loc[movies.num_voted_users > 25000 ]).sort_values(by='imdb_score' , ascending=False).head(250)
IMDb_Top_250

director_name  num_critic_for_reviews       gross  \
1937        Frank Darabont                   199.0   28.341469   
3466  Francis Ford Coppola                   208.0  134.821952   
2837  Francis Ford Coppola                   149.0   57.300000   
66       Christopher Nolan                   645.0  533.316061   
339          Peter Jackson                   328.0  377.019252   
1874      Steven Spielberg                   174.0   96.067179   
4498          Sergio Leone                   181.0    6.100000   
3355     Quentin Tarantino                   215.0  107.930000   
97       Christopher Nolan                   642.0  292.568851   
683          David Fincher                   315.0   37.023395   
2051        Irvin Kershner                   223.0  290.158751   
836        Robert Zemeckis                   149.0  329.691196   
270          Peter Jackson                   297.0  313.837577   
3867          Milos Forman                   149.0  112.000000   
4029    Fernando Meirelles                   214.0    7.563397   
3024          George Lucas                   282.0  460.935665   
4747        Akira Kurosawa                   153.0    0.269061   
340          Peter Jackson                   294.0  340.478898   
1903       Martin Scorsese                   192.0   46.836394   
654         Lana Wachowski                   313.0  171.383253   
2158        Jonathan Demme                   185.0  130.727000   
4427       Charles Chaplin                   120.0    0.163245   
3175             Tony Kaye                   162.0    6.712241   
648       Steven Spielberg                   219.0  216.119491   
3592          Bryan Singer                   162.0   23.272306   
2373        Hayao Miyazaki                   246.0   10.049886   
1600         David Fincher                   216.0  100.125340   
96       Christopher Nolan                   712.0  187.991439   
283           Ridley Scott                   265.0  187.670866   
1571  Francis Ford Coppola                   261.0   78.800000   
...                    ...                     ...         ...   
353          John Lasseter                   191.0  245.823397   
1807           John Landis                   125.0   54.200000   
3357            Dan Gilroy                   534.0   32.279955   
2418          Edward Zwick                    60.0   26.830000   
1735         James Mangold                   291.0  119.518352   
1713          Edward Zwick                    60.0   26.830000   
3510           Yash Chopra                    29.0    2.921738   
69             Jon Favreau                   486.0  318.298180   
0            James Cameron                   723.0  760.505847   
3452           Paul Haggis                   287.0   54.557348   
927         Andrew Adamson                   212.0  267.652016   
162           Dean DeBlois                   292.0  176.997107   
3080      Steven Spielberg                   215.0  434.949459   
2449            Tim Burton                   148.0    5.887457   
3768          Duncan Jones                   415.0    5.009677   
1871          Pierre Morel                   309.0  145.000989   
99           Peter Jackson                   645.0  303.001229   
4415      Fabián Bielinsky                    94.0    1.221261   
23           Peter Jackson                   509.0  258.355354   
2177            Tim Burton                   111.0   56.362352   
3677  Christophe Barratier                   112.0    3.629758   
2493           Yimou Zhang                   283.0    0.084961   
1748          F. Gary Gray                   349.0  161.029270   
3595      Darren Aronofsky                   391.0   26.236603   
2558          Edgar Wright                   285.0   23.618786   
4640       Cristian Mungiu                   233.0    1.185783   
2492        John Carpenter                   318.0   47.000000   
4821        John Carpenter                   318.0   47.000000   
639           Michael Mann                   209.0   

In [39]:
# Using rank() function, ranks are generate to the 250 rows in IMDb_Top_250 dataframe and assigned to a new column 'rank'
# Here, the method used for generating rank is 'first', meaning the ranks are ordered in the order they appear in the dataframe
# By default the ranks assigned will be of float type. Applying astype(int) to converrt the ranks to integers.

IMDb_Top_250['rank'] = IMDb_Top_250['imdb_score'].rank(ascending=False,method='first').astype('int')

# Ranking method ='first' based on clarification by TA
# To provide Upgrad Leaderboard like ranking, method='min' can be used as below
# IMDb_Top_250['rank'] = IMDb_Top_250['imdb_score'].rank(ascending=False,method='min').astype('int') 

In [40]:
# Print first few rows of the dataframe IMDb_Top_250 to check sorting and creation of rank
IMDb_Top_250.head(10)

director_name  num_critic_for_reviews       gross  \
1937        Frank Darabont                   199.0   28.341469   
3466  Francis Ford Coppola                   208.0  134.821952   
2837  Francis Ford Coppola                   149.0   57.300000   
66       Christopher Nolan                   645.0  533.316061   
339          Peter Jackson                   328.0  377.019252   
1874      Steven Spielberg                   174.0   96.067179   
4498          Sergio Leone                   181.0    6.100000   
3355     Quentin Tarantino                   215.0  107.930000   
97       Christopher Nolan                   642.0  292.568851   
683          David Fincher                   315.0   37.023395   

                                genres       actor_1_name  \
1937                       Crime|Drama     Morgan Freeman   
3466                       Crime|Drama          Al Pacino   
2837                       Crime|Drama     Robert De Niro   
66         Action|Crime|Drama|Thriller     Christian Bale   
339     Action|Adventure|Drama|Fantasy      Orlando Bloom   
1874           Biography|Drama|History        Liam Neeson   
4498                           Western     Clint Eastwood   
3355                       Crime|Drama       Bruce Willis   
97    Action|Adventure|Sci-Fi|Thriller  Leonardo DiCaprio   
683                              Drama          Brad Pitt   

                                         movie_title  num_voted_users  \
1937                       The Shawshank Redemption           1689764   
3466                                  The Godfather           1155770   
2837                         The Godfather: Part II            790926   
66                                  The Dark Knight           1676169   
339   The Lord of the Rings: The Return of the King           1215718   
1874                               Schindler's List            865020   
4498                 The Good, the Bad and the Ugly            503509   
3355                                   Pulp Fiction           1324680   
97                                        Inception           1468200   
683                                      Fight Club           1347461   

      num_user_for_reviews language  budget  title_year  imdb_score  \
1937                4144.0  English    25.0      1994.0         9.3   
3466                2238.0  English     6.0      1972.0         9.2   
2837                 650.0  English    13.0      1974.0         9.0   
66                  4667.0  English   185.0      2008.0         9.0   
339                 3189.0  English    94.0      2003.0         8.9   
1874                1273.0  English    22.0      1993.0         8.9   
4498                 780.0  Italian     1.2      1966.0         8.9   
3355                2195.0  English     8.0      1994.0         8.9   
97                  2803.0  English   160.0      2010.0         8.8   
683                 2968.0  English    63.0      1999.0         8.8   

      movie_facebook_likes      profit  rank  
1937                108000    3.341469     1  
3466                 43000  128.821952     2  
2837                 14000   44.300000     3  
66                   37000  348.316061     4  
339                  16000  283.019252     5  
1874                 41000   74.067179     6  
4498                 20000    4.900000     7  
3355                 45000   99.930000     8  
97                  175000  132.568851     9  
683                  48000  -25.976605    10

In [41]:
# Write your code to extract top foreign language films from 'IMDb_Top_250' here
# [IMDb_Top_250.language != 'English'] : selects all the rows for which the language is not(!) 'English'
Top_Foreign_Lang_Film = IMDb_Top_250.loc[IMDb_Top_250.language != 'English']

In [42]:
# Print first few rows of Top_Foreign_Lang_Film dataframe to confirm that only non-English movies have been chosen
Top_Foreign_Lang_Film.head(10)

director_name  num_critic_for_reviews      gross  \
4498                      Sergio Leone                   181.0   6.100000   
4029                Fernando Meirelles                   214.0   7.563397   
4747                    Akira Kurosawa                   153.0   0.269061   
2373                    Hayao Miyazaki                   246.0  10.049886   
4921                      Majid Majidi                    46.0   0.925402   
4259  Florian Henckel von Donnersmarck                   215.0  11.284657   
1329                    S.S. Rajamouli                    44.0   6.498000   
4659                    Asghar Farhadi                   354.0   7.098492   
1298                Jean-Pierre Jeunet                   242.0  33.201661   
4105                    Chan-wook Park                   305.0   2.181290   

                                  genres       actor_1_name  \
4498                             Western     Clint Eastwood   
4029                         Crime|Drama        Alice Braga   
4747              Action|Adventure|Drama    Takashi Shimura   
2373  Adventure|Animation|Family|Fantasy     Bunta Sugawara   
4921                        Drama|Family     Bahare Seddiqi   
4259                      Drama|Thriller     Sebastian Koch   
1329  Action|Adventure|Drama|Fantasy|War   Tamannaah Bhatia   
4659                       Drama|Mystery    Shahab Hosseini   
1298                      Comedy|Romance  Mathieu Kassovitz   
4105              Drama|Mystery|Thriller       Min-sik Choi   

                          movie_title  num_voted_users  num_user_for_reviews  \
4498  The Good, the Bad and the Ugly            503509                 780.0   
4029                     City of God            533200                 749.0   
4747                   Seven Samurai            229012                 596.0   
2373                   Spirited Away            417971                 902.0   
4921              Children of Heaven             27882                 130.0   
4259             The Lives of Others            259379                 407.0   
1329        Baahubali: The Beginning             62756                 410.0   
4659                    A Separation            151812                 264.0   
1298                          Amélie            534262                1314.0   
4105                          Oldboy            356181                 809.0   

        language     budget  title_year  imdb_score  movie_facebook_likes  \
4498     Italian   1.200000      1966.0         8.9                 20000   
4029  Portuguese   3.300000      2002.0         8.7                 28000   
4747    Japanese   2.000000      1954.0         8.7                 11000   
2373    Japanese  19.000000      2001.0         8.6                 28000   
4921     Persian   0.180000      1997.0         8.5                     0   
4259      German   2.000000      2006.0         8.5                 39000   
1329      Telugu  18.026148      2015.0         8.4                 21000   
4659     Persian   0.500000      2011.0         8.4                 48000   
1298      French  77.000000      2001.0         8.4                 39000   
4105      Korean   3.000000      2003.0         8.4                 43000   

         profit  rank  
4498   4.900000     7  
4029   4.263397    15  
4747  -1.730939    17  
2373  -8.950114    26  
4921   0.745402    43  
4259   9.284657    46  
1329 -11.528148    47  
4659   6.598492    49  
1298 -43.798339    52  
4105  -0.818710    57

In [43]:
# Check whether 'Veer-Zaara' is present in Top_Foreign_Lang_Film
Top_Foreign_Lang_Film[Top_Foreign_Lang_Film['movie_title'].str.contains('Veer-Zaara')]

director_name  num_critic_for_reviews     gross                 genres  \
3510   Yash Chopra                    29.0  2.921738  Drama|Musical|Romance   

        actor_1_name  movie_title  num_voted_users  num_user_for_reviews  \
3510  Shah Rukh Khan  Veer-Zaara             34449                 119.0   

     language  budget  title_year  imdb_score  movie_facebook_likes    profit  \
3510    Hindi     7.0      2004.0         7.9                  2000 -4.078262   

      rank  
3510   227

**Checkpoint 3:** Can you spot `Veer-Zaara` in the dataframe? 

- ### Subtask 3.5: Find the best directors

    1. Group the dataframe using the `director_name` column.
    2. Find out the top 10 directors for whom the mean of `imdb_score` is the highest and store them in a new dataframe `top10director`. 

In [44]:
# Write your code for extracting the top 10 directors here

# Group the 'movies' dataframe by 'director_name' column using group_by() function
movies_by_director = movies.groupby('director_name')

# Get Top 10 Directors by mean of imdb_score
# movies_by_director['imdb_score'].mean() : calculates mean of 'imdb_score' column grouped by 'director_name' column
# sort_values(ascending=False) : sorts the resultant series in descending order
# [:10] : select only the first 10 rows

top10director = movies_by_director['imdb_score'].mean().sort_values(ascending=False)[:10]

#Convert 'top10director' series to dataframe , reset index and give meaningful column names
top10director = top10director.to_frame().reset_index().\
                      rename(columns={'director_name':'Dicector_Name','imdb_score':'Avg_IMDB_Score'})

# Print the dataframe 'top10director' to confirm presense of 'Damien Chazelle'
top10director 

Dicector_Name  Avg_IMDB_Score
0    Charles Chaplin        8.600000
1          Tony Kaye        8.600000
2         Ron Fricke        8.500000
3    Damien Chazelle        8.500000
4       Majid Majidi        8.500000
5   Alfred Hitchcock        8.500000
6       Sergio Leone        8.433333
7  Christopher Nolan        8.425000
8     Asghar Farhadi        8.400000
9   Richard Marquand        8.400000

**Checkpoint 4:** No surprises that `Damien Chazelle` (director of Whiplash and La La Land) is in this list.

-  ### Subtask 3.6: Find popular genres

You might have noticed the `genres` column in the dataframe with all the genres of the movies seperated by a pipe (`|`). Out of all the movie genres, the first two are most significant for any film.

1. Extract the first two genres from the `genres` column and store them in two new columns: `genre_1` and `genre_2`. Some of the movies might have only one genre. In such cases, extract the single genre into both the columns, i.e. for such movies the `genre_2` will be the same as `genre_1`.
2. Group the dataframe using `genre_1` as the primary column and `genre_2` as the secondary column.
3. Find out the 5 most popular combo of genres by finding the mean of the gross values using the `gross` column and store them in a new dataframe named `PopGenre`.

In [45]:
# Write your code for extracting the first two genres of each movie here

# Function to extract the first 2 genre
def genres_1_2(genres):
    '''genres_1_2(genres) : function to extract the first two genres
       Input : movies.genres series from movies dataframe
       Return : A series with a list containing two values
              : First values corresponding to 'genre_1' and second one to 'genre_2'
    '''
    # Since the 'genres' series is of type 'object', it is represented as a string.
    # Split the input 'genres' series on "|" symbol using split('|')
    # This results in a list of elements each representing the various genre mentioned in 'genres' series.
    list_genres=genres.split('|')
    
    # Check the length of 'list_genres'. 
    if len(list_genres) > 1:                              # Movie has more than 1 genre identified against it.
        return pd.Series([list_genres[0], list_genres[1]])
    else:                                                 # Movie has only 1 genre which needs to be propogated to 'genre_2'.
        return pd.Series([list_genres[0], list_genres[0]])

movies[['genre_1','genre_2']] = movies['genres'].apply(genres_1_2)

In [46]:
# Display head of 'movies' dataframe to confirm inclusion of columns 'genre_1' and 'genre_2'.
movies.head()

director_name  num_critic_for_reviews       gross  \
0        James Cameron                   723.0  760.505847   
29     Colin Trevorrow                   644.0  652.177271   
26       James Cameron                   315.0  658.672302   
3024      George Lucas                   282.0  460.935665   
3080  Steven Spielberg                   215.0  434.949459   

                                genres         actor_1_name  \
0      Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
29    Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
26                       Drama|Romance    Leonardo DiCaprio   
3024   Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
3080                     Family|Sci-Fi         Henry Thomas   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
29                        Jurassic World            418214   
26                               Titanic            793059   
3024  Star Wars: Episode IV - A New Hope            911097   
3080          E.T. the Extra-Terrestrial            281842   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
29                  1290.0  English   150.0      2015.0         7.0   
26                  2528.0  English   200.0      1997.0         7.7   
3024                1470.0  English    11.0      1977.0         8.7   
3080                 515.0  English    10.5      1982.0         7.9   

      movie_facebook_likes      profit genre_1    genre_2  
0                    33000  523.505847  Action  Adventure  
29                  150000  502.177271  Action  Adventure  
26                   26000  458.672302   Drama    Romance  
3024                 33000  449.935665  Action  Adventure  
3080                 34000  424.449459  Family     Sci-Fi

In [47]:
# Write your code for grouping the dataframe here
# Group the 'movies' dataframe using 'genre_1' and primary and 'genre_2' as secondary grouping column using group_by() function.
movies_by_genre = movies.groupby(['genre_1', 'genre_2'])

In [48]:
# Write your code for getting the 5 most popular combo of genres here
# movies_by_genre.gross.mean() : Find mean of 'gross' column grouped by 'genre_1' and 'genre_2'
# sort_values(ascending = False) : sort the resultant series in descending order
# [:5] : Selects only first 5 rows.
PopGenre = movies_by_genre.gross.mean().sort_values(ascending = False)[:5]

#Convert 'PopGenre' series to dataframe , reset index and give meaningful column names
PopGenre = PopGenre.to_frame().reset_index().\
                      rename(columns={'genre_1':'Genre_1','genre_2':'Genre_2','gross':'Avg_Gross_Collection'})

# Display 'PopGenre' dataframe and check which is the most popular genre combo
PopGenre

Genre_1    Genre_2  Avg_Gross_Collection
0     Family     Sci-Fi            434.949459
1  Adventure     Sci-Fi            228.627758
2  Adventure     Family            118.919540
3  Adventure  Animation            116.998550
4     Action  Adventure            109.595465

**Checkpoint 5:** Well, as it turns out. `Family + Sci-Fi` is the most popular combo of genres out there! 

-  ### Subtask 3.7: Find the critic-favorite and audience-favorite actors

    1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
    2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
    3. Group the combined dataframe using the `actor_1_name` column.
    4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [49]:
# Write your code for creating three new dataframes here

# Include all movies in which Meryl_Streep is the lead
# Using isin() function , select only rows where 'actor_1_name' is 'Meryl Streep'
Meryl_Streep = movies.loc[movies['actor_1_name'].isin(['Meryl Streep']), :]

In [50]:
# Include all movies in which Leo_Caprio is the lead
# Using isin() function , select only rows where 'actor_1_name' is 'Leonardo DiCaprio'
Leo_Caprio = movies.loc[movies['actor_1_name'].isin(['Leonardo DiCaprio']), :]

In [51]:
# Include all movies in which Brad_Pitt is the lead
# Using isin() function , select only rows where 'actor_1_name' is 'Brad Pitt'
Brad_Pitt = movies.loc[movies['actor_1_name'].isin(['Brad Pitt']), :]

In [52]:
# Write your code for combining the three dataframes here
# Concat the three dataframes by passing them as a list to pd.concat() function
Combined = pd.concat([Meryl_Streep,Leo_Caprio,Brad_Pitt])

In [53]:
# Create a DataFrameGroupBy object by grouping on 'actor_1_name' column in 'Combined' dataframe.
by_actor_1_name = Combined.groupby('actor_1_name')

In [54]:
# Write the code for finding the mean of critic reviews and audience reviews here

# Actor by Critic Review
# by_actor_1_name.num_critic_for_reviews.mean() : Find mean of 'num_critic_for_reviews' column grouped by 'actor_1_name'
# sort_values(ascending = False) : sort the resultant series in descending order
actor_critic_review = by_actor_1_name.num_critic_for_reviews.mean().sort_values(ascending = False)

#Convert 'actor_critic_review' series to dataframe , reset index and give meaningful column names
actor_critic_review = actor_critic_review.to_frame().reset_index().\
                      rename(columns={'actor_1_name':'Actor_Name','num_critic_for_reviews':'Avg_Critic_Review'})
#-------------------------------------------------------------------------------------------------------------------#
# Actor by Audience Review
# by_actor_1_name.num_user_for_reviews.mean() : Find mean of 'num_user_for_reviews' column grouped by 'actor_1_name'
# sort_values(ascending = False) : sort the resultant series in descending order
actor_audience_review = by_actor_1_name.num_user_for_reviews.mean().sort_values(ascending = False)

#Convert 'actor_audience_review' series to dataframe , reset index and give meaningful column names
actor_audience_review = actor_audience_review.to_frame().reset_index().\
                        rename(columns={'actor_1_name':'Actor_Name','num_user_for_reviews':'Avg_Audience_Review'})

In [55]:
# Print the dataframes 'actor_critic_review' and 'actor_audience_review'
# Check if 'Leonardo DiCaprio' tops both the lists
# Actor by Critic Review
actor_critic_review

Actor_Name  Avg_Critic_Review
0  Leonardo DiCaprio         330.190476
1          Brad Pitt         245.000000
2       Meryl Streep         181.454545

In [56]:
# Actor by Audience Review
actor_audience_review

Actor_Name  Avg_Audience_Review
0  Leonardo DiCaprio           914.476190
1          Brad Pitt           742.352941
2       Meryl Streep           297.181818

**Checkpoint 6:** `Leonardo` has aced both the lists! 